### PR TITLE
Add tempest-k8s into openstack terraform plan

### PR DESCRIPTION
**Pull request by chanchiwai-ray**
This PR adds tempest-k8s into openstack terraform plan.

- The tempest-k8s will only be deployed if the validation plugin is enabled
- The integration to grafana-agent-k8s (loki and grafana) will only be established if the observability plugin is enabled

Relates to:  https://github.com/agileshaw/snap-openstack/pull/2

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -565,7 +565,7 @@ resource "juju_application" "openstack-exporter" {
   charm {
     name     = "openstack-exporter-k8s"
     channel  = var.openstack-exporter-channel == null ? var.openstack-channel : var.openstack-exporter-channel
-    revision = var.openstack-exporter-channel
+    revision = var.openstack-exporter-revision
```
Comment (samuelallan72) [inline main.tf:568]: This doesn't seem related here?  If it's fixing an unrelated bug in existing code, maybe submit it separated directly to upstream?

File: variables.tf [hunk 2]
```diff
@@ -636,3 +636,51 @@ variable "horizon-plugins" {
...
+  description = "Operator channel for Tempest deployment"
+  type        = string
+  default     = null
+}
+
+variable "tempest-revision" {
+  description = "Operator channel revision for Tempest deployment"
+  type        = number
+  default     = null
+}
```
Comment (samuelallan72) [inline variables.tf:656]: why do we need both tempest-channel and tempest-revision?  Also, can we make it clearer that this is config for the tempest _charm_?  Or is `Operator` standard language here to denote it's for a charm?

Perhaps a variable named like `charm-tempest-channel` or `tempest-k8s-channel` might make it clearer that it's for the charm, not the tempest snap or tempest itself.

Comment (samuelallan72) [reply variables.tf:656]: @chanchiwai-ray `tempest-revision` doesn't seem to be used anywhere?

**Review by samuelallan72 — CHANGES_REQUESTED** [review]
Mostly looks fine, but I had a couple of questions. :)

File: variables.tf (hunk 2 above)
Comment (agileshaw) [reply variables.tf:656]: I'm ok with `tempest-revision` being added here (and also the name), which enables the user to pin-point the revision they wants. I imagine this to be especially useful when sunbeam starts to support manifest.

File: main.tf (hunk 1 above)
Comment (agileshaw) [reply main.tf:568]: +1 on not including changes to opentack-exporter here. We can handle this change together with openstack-exporter integration with grafana-agent-k8s when we get there.

File: main.tf [hunk 3]
```diff
@@ -866,3 +866,63 @@ resource "juju_integration" "ldap-to-keystone" {
...
 }
+
+resource "juju_application" "tempest" {
+  count = var.enable-validation ? 1 : 0
+  name  = "tempest"
+  model = juju_model.sunbeam.name
+
+  charm {
+    name     = "tempest-k8s"
+    channel  = var.tempest-channel
```
Comment (agileshaw) [inline main.tf]: Like other charms defined, I think we should set the default channel to openstack channel if not specified?
```suggestion
    channel = var.tempest-channel == null ? var.openstack-channel : var.tempest-channel
```

File: main.tf [hunk 4]
```diff
@@ -866,3 +866,63 @@ resource "juju_integration" "ldap-to-keystone" {
...
+    revision = var.tempest-revision
+  }
+
+  units  = 1
```
Comment (agileshaw) [inline main.tf:881]: How is `count` preventing deployment when unit is always set to 1? I didn't find `count` in the source code of juju terraform provider but saw it being used in integrations (`count.index`), so not sure how exactly it is taking effect.

**Review by agileshaw — CHANGES_REQUESTED** [review]

File: main.tf (hunk 4 above)
Comment (rgildein) [reply main.tf:881]: The count is coming from Terraform itself, so any resource can have it, and it means number of instances. That means if it's 0 no tempest applications would be deployed, and it's 1 you will get 1 tempest application, which is defined with single unit (coming from Juju Provider as you mentioned).

I think that if we use count: 2 it should fail, because you cannot have two application with same name on Juju.

File: variables.tf [hunk 5]
```diff
@@ -636,3 +636,51 @@ variable "horizon-plugins" {
...
+
+variable "tempest-config" {
+  description = "Operator config for Tempest deployment"
+  type        = map(string)
+  default     = {}
+}
+
+variable "grafana-agent-k8s-name" {
```
Comment (rgildein) [inline variables.tf:664]: If I'm correct, the Tempest will be deployed on its own model, or not? If yes, for what this will be used?

File: variables.tf [hunk 6]
```diff
@@ -636,3 +636,51 @@ variable "horizon-plugins" {
...
+  description = "Operator name for Grafana Agent K8s deployment"
+  type        = string
+  default     = "grafana-agent-k8s"
+}
+
+variable "metrics-endpoint" {
```
Comment (rgildein) [inline variables.tf:670]: I do not understand this (and 2 following) variable either, it can be used when you want to use your own cos instead of cos deployed with cos plugin?

**Review by rgildein — COMMENTED** [review]
Only few small comments/questions, otherwise it looks good to me.

File: main.tf (hunk 4 above)
Comment (agileshaw) [reply main.tf:881]: Aha I see. Yea, just searched it up and saw that `count` is a meta argument of terraform itself. Thanks.

File: main.tf (hunk 1 above)
Comment (chanchiwai-ray) [reply main.tf:568]: I fixed it already in the upstream.

File: variables.tf (hunk 2 above)
Comment (chanchiwai-ray) [reply variables.tf:656]: It was their practice to ensure variables has all channel, revision, and config: https://github.com/canonical/sunbeam-terraform/commit/5a6494b4f8e1c16f127c40cb47f7f9fb11ec7461

File: variables.tf (hunk 5 above)
Comment (chanchiwai-ray) [reply variables.tf:664]: Grafana agent k8s and tempest-k8s will be on the same model: the openstack model (i.e. the control plane). But they will be deployed using different terraform plans. In the terraform plan with tempest-k8s, we don't have access to the name of the other juju resource in another terraform plan (I am not an expert to terraform as well, maybe there is a way? using output.tf), so we need to pass the name here. It is used here https://github.com/agileshaw/sunbeam-terraform/pull/1/files/5c615cb9b146fc3271cefed06a10fb2f6ef1bed1#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR910

File: variables.tf (hunk 6 above)
Comment (chanchiwai-ray) [reply variables.tf:670]: The relations are not related to COS; they are `workload charms (openstack model) <--**this one**--> grafana-agent-k8s (openstack model) <----> COS (non-openstack model) `

These are some variable definitions (grafana-agent-k8s-name, metrics-endpoint, logging-provider, grafana-dashboards-consumer) to be used by any application in the openstack model. For example, the openstack-exporter (which is currently directly connected to COS model, and we might need to fix it in the later EPIC), and the tempest-k8s. 

The example usage is:
1. when enabling the observability plugin, we fill the [variables ](https://github.com/agileshaw/sunbeam-terraform/pull/1/files/5c615cb9b146fc3271cefed06a10fb2f6ef1bed1#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR900-R928) with appropriate values
2. these values will be updated in the actual terraform plan for the openstack model (i.e. the main.tf)
3. observability plugin will apply the plan with the updated value to build the relation between grafana-agent-k8s and tempest-k8s (and in the later EPIC, also the openstack-exporter-k8s)

File: main.tf (hunk 3 above)
Comment (chanchiwai-ray) [reply main.tf]: Done

**Review by agileshaw — APPROVED** [review]
LGTM

File: variables.tf (hunk 5 above)
Comment (rgildein) [reply variables.tf:664]: Link you provided is not working.
I know that tempest-k8s and grafana-agent-k8s will be on same model, but I thought that it will be separate model, not the control plane one. So now I want to understand why we need/want to have different name for grafana-agent-k8s. We will use same (for multiple apps) grafana-agent? Or we will deploy one only for tempest-k8s?

File: variables.tf (hunk 6 above)
Comment (rgildein) [reply variables.tf:670]: I know they are not related to COS directly, but I dot understand why you need them instead of using name define in proper .tf file? Why this needs to be configurable?
Not an expert, just to clarify thinks.
BTW, your link is not working again.

File: main.tf (hunk 1 above)
Comment (samuelallan72) [reply main.tf:568]: Perfect, could you rebase on upstream so it's clear here. :)

File: variables.tf (hunk 5 above)
Comment (chanchiwai-ray) [reply variables.tf:664]: We could hardcode the name here, and make sure the name defined in observability plugin matches the name defined here. But there will be two degree-of-freedoms: (1) grafana-agent-k8s name defined here, and (2) grafana-agent-k8s name defined in observability plugin, and (1) and (2) must match.

Technically, the name of grafana-agent-k8s is unknown to "this" terraform plan, so I make sense to make it a variable

File: variables.tf (hunk 6 above)
Comment (chanchiwai-ray) [reply variables.tf:670]: They need to be added / removed dynamically. You cannot hardcode them

File: variables.tf (hunk 5 above)
Comment (chanchiwai-ray) [reply variables.tf:664]: But yes, my argument here is quite weak. We could hardcode the name if you prefer

File: main.tf (hunk 1 above)
Comment (chanchiwai-ray) [reply main.tf:568]: cc @agileshaw , your main branch is not up-to-date

File: variables.tf (hunk 5 above)
Comment (rgildein) [reply variables.tf:664]: Ok, so they are or two different repos. Now it makes sense, but it's still have felling that it's strange.

File: variables.tf (hunk 6 above)
Comment (rgildein) [reply variables.tf:670]: Again reason is that we have two repos for final Terraform plan, which is strange, but now I understand why we needed it.

**Review by rgildein — APPROVED** [review]
